src/Cryptol/TypeCheck/Subst.hs

```diff
@@ -67,18 +77,39 @@ emptySubst =
     , suDefaulting = False
     }
 
-singleSubst :: TVar -> Type -> Subst
-singleSubst v@(TVFree i _ _tps _) t =
+-- | Reasons to reject a single-variable substitution.
+data SubstError
+  = SubstRecursive
+  -- ^ 'TVar' maps to a type containing the same variable.
+  | SubstEscaped [TParam]
+  -- ^ 'TVar' maps to a type containing one or more out-of-scope bound variables.
+
+singleSubst :: TVar -> Type -> Either SubstError Subst
+singleSubst x t
```

Comment:
We should probably also check that the kinds of the variable and the type are the same here, and report the corresponding error if not.

Reply:
That's a good idea. I'll add an extra constructor to `SubstError` for kind mismatch, add kind-correctness as a third documented invariant for the `Subst` datatype, and add a kind check to `singleSubst`. This will let us simplify the definition of `bindVar` too.